favor/favor.py

```diff
@@ -87,20 +88,35 @@ def phi(x):
         # (batch, n, r)
         phi_q = phi(queries)
 
-        out = torch.matmul( # (batch, n, dim)
-            phi_q,
-            torch.matmul( # (batch, r, dim)
-                phi_k.permute(0, 2, 1), values
+        if self.causal:
+            # outer products of keys and values: (batch, n, r, dim)
+            k_v_prod = torch.matmul(phi_k[:, :, :, None], values[:, :, None, :])
+
+            out = torch.matmul(         # (batch, n, dim)
+                phi_q[:, :, None, :],   # (batch, n, 1, r)
+                k_v_prod.cumsum(dim=1)  # (batch, n, r, dim)
+            ).squeeze(2)
+
+            # normalization factors: (batch, n, 1)
+            norm = torch.matmul(
+                phi_q[:, :, None, :],           # (batch, n, 1, r)
+                phi_k.cumsum(dim=1)[..., None]  # (batch, n, r, 1)
+            ).squeeze(2)
+        else:
+            out = torch.matmul( # (batch, n, dim)
+                phi_q,
+                torch.matmul( # (batch, r, dim)
+                    phi_k.permute(0, 2, 1), values
+                )
             )
-        )
 
-        # rescaling: (batch, n, 1)
-        scale = torch.matmul(
-            phi_q,
-            torch.sum(phi_k, dim=1)[..., None] # (batch, r, 1)
-        ) 
+            # normalization factors: (batch, n, 1)
+            norm = torch.matmul(
+                phi_q,
+                phi_k.sum(dim=1)[..., None] # (batch, r, 1)
+            )
 
-        out = out / scale
+        out /= norm
```

Comment:
no, be careful that this probably does not propagate gradient: we should not do in place operations. or has this been fixed in latest pytorch ?

Reply:
I remember reading [this](https://pytorch.org/docs/stable/notes/autograd.html#in-place-correctness-checks):
> [...] if you’re using in-place functions and not seeing any errors, you can be sure that the computed gradients are correct.

So that means it should be safe, no? (unless there's a bug I don't know about)

Reply:
please revert just this and then I'll merge
I really encountered errors due to that in the past.

_or_: really test backprop through this, not only forward

Reply:
OK!